src/DataStructures/ComplexDataVector.hpp

```diff
@@ -0,0 +1,246 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class Data.
+
+#pragma once
+
+#include <complex>
+#include <cstddef>
+#include <functional>  // for std::reference_wrapper
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/VectorImpl.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/StdArrayHelpers.hpp"
+
+// IWYU pragma: no_forward_declare ConstantExpressions_detail::pow
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief Stores a collection of complex function values.
+ *
+ * \details Use DataVector to represent function values on the computational
+ * domain. Note that interpreting the data also requires knowledge of the points
+ * that these function values correspond to.
+ *
+ * A ComplexDataVector holds an array of contiguous data. The ComplexDataVector
+ * can be owning, meaning the array is deleted when the ComplexDataVector goes
+ * out of scope, or non-owning, meaning it just has a pointer to an array.
+ *
+ * Refer to the \ref DataStructuresGroup documentation for a list of other
+ * available types.
+ *
+ * ComplexDataVectors support a variety of mathematical operations that are
+ * applicable to nodal coefficients. In addition to common arithmetic operations
+ * such as elementwise addition, subtraction, multiplication and division the
+ * following elementwise operations are implemented:
+ *
+ * - abs  (returns DataVector, as is real)
+ * - acos
+ * - acosh
+ * - asin
+ * - asinh
+ * - atan
+ * - atan2
+ * - atanh
+ * - cos
+ * - cosh
+ * - erf
+ * - erfc
+ * - exp
+ * - exp10
+ * - imag  (returns DataVector, as is real)
+ * - log
+ * - log10
+ * - max
+ * - min
+ * - pow
+ * - real  (returns DataVector, as is real)
+ * - sin
+ * - sinh
+ * - sqrt
+ * - tan
+ * - tanh
+ */
+class ComplexDataVector : public detail::VectorImpl<std::complex<double>,
+                                                    ComplexDataVector>
+{
+public:
+  using detail::VectorImpl<std::complex<double>,ComplexDataVector>::operator=;
+  using detail::VectorImpl<std::complex<double>,ComplexDataVector>::VectorImpl;
+
+  MAKE_EXPRESSION_MATH_ASSIGN_PV(+=, ComplexDataVector)
+  MAKE_EXPRESSION_MATH_ASSIGN_PV(-=, ComplexDataVector)
+  MAKE_EXPRESSION_MATH_ASSIGN_PV(*=, ComplexDataVector)
+  MAKE_EXPRESSION_MATH_ASSIGN_PV(/=, ComplexDataVector)
+};
+
+/*
+ * \brief evaluate the absolute value of a ComplexDataVector, returning a
+ * DataVector
+ */
+DataVector abs(const ComplexDataVector& t) noexcept;
+
+DataVector real(const ComplexDataVector& t) noexcept;
+
+DataVector imag(const ComplexDataVector& t) noexcept;
+
+ComplexDataVector conj(const ComplexDataVector& t) noexcept;
+
+namespace blaze {
+  DATAVECTOR_BLAZE_TRAIT_SPEC(ComplexDataVector)
```

Comment:
Please figure out how to get your editor to do the right formatting.

Reply:
Sorry. Done (both formatting this and setting emacs).